pkg/cosmos/txmgr/models.go

```diff
@@ -0,0 +1,107 @@
+package txmgr
+
+import (
+	"bytes"
+
+	wasmtypes "github.com/CosmWasm/wasmd/x/wasm/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/bank/types"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/pkg/errors"
+
+	"github.com/smartcontractkit/chainlink-cosmos/pkg/cosmos"
+
+	txmgrtypes "github.com/smartcontractkit/chainlink-relay/pkg/txmgr/types"
+)
+
+// Type aliases for Cosmos specific implementation of generic txmgr
+type (
+	TxRequest = txmgrtypes.TxRequest[Address, TxHash]
+)
+
+var (
+	typeMsgSend            = sdk.MsgTypeURL(&types.MsgSend{})
+	typeMsgExecuteContract = sdk.MsgTypeURL(&wasmtypes.MsgExecuteContract{})
+)
+
+// Encodes cosmos native txn type into bytes for generic txmgr
+func EncodePayload(msg sdk.Msg) ([]byte, error) {
+	switch ms := msg.(type) {
+	case *wasmtypes.MsgExecuteContract:
+		_, err := sdk.AccAddressFromBech32(ms.Sender)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to encode payload as parse sender failed: %s", ms.Sender)
+		}
+
+	case *types.MsgSend:
+		_, err := sdk.AccAddressFromBech32(ms.FromAddress)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to encode payload as parse sender failed: %s", ms.FromAddress)
+		}
+
+	default:
+		return nil, &cosmos.ErrMsgUnsupported{Msg: msg}
+	}
+	raw, err := proto.Marshal(msg)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to encode payload as proto marshal failed: %s", msg)
+	}
+	return raw, nil
+}
+
+func DecodePayload(msgType string, raw []byte) (msg sdk.Msg, err error) {
+	switch msgType {
+	case typeMsgSend:
+		var ms types.MsgSend
+		err := ms.Unmarshal(raw)
+		if err != nil {
+			return nil, err
+		}
+		return &ms, nil
+	case typeMsgExecuteContract:
+		var ms wasmtypes.MsgExecuteContract
+		err := ms.Unmarshal(raw)
+		if err != nil {
+			return nil, err
+		}
+		return &ms, nil
+	}
+	return nil, errors.Errorf("unrecognized message type: %s", msgType)
+}
+
+// A wrapper for sdk.Address. Uses fixed max address size of 32-byte size to satisfy comparable constraint in Hashable interface.
+// Cosmos addresses shorter than 32 bytes are zero-padded.
+// https://docs.cosmos.network/v0.46/basics/accounts.html#keys-accounts-addresses-and-signatures
+type Address [32]byte
+
+// Create a new generic txmgr Address from a sdk.Address
+func NewAddress(addr sdk.AccAddress) Address {
+	var a Address
+	copy(a[:], addr[:])
+	return a
+}
+
+// Convert a generic txmgr Address to a Cosmos sdk.AccAddress
+func ToCosmosAddress(addr Address) sdk.AccAddress {
+	return sdk.AccAddress(bytes.TrimRight(addr[:], "\x00"))
```

Comment:
If I understand this correctly, does this remove all trailing 0s from the address?

What if the actual cosmos address had 0s in the address?
Or, I assume cosmos address wouldn't have 0s(unicode 0s, not string of "0") in the address, so its safe to do this?

Reply:
yup, this will remove trailing zeros from the addresses. 

> What if the actual cosmos address had 0s in the address?

The cosmos addresses can have 0s in the address, but it would not have unicode 0s

Cosmos addresses are derived from public keys using the Bech32 encoding scheme. Because Bech32 uses a specific set of 32 characters, the Unicode null character (U+0000).